Completed Tasks 4 & 5 with unit tests.

FormalProcess3909 opened the pull request:
Completed Tasks 4 & 5 with unit tests. Utilized the utils.test.ts to ensure tasks 4 & 5 unit tests ran correctly. Created a separate utility file for creating JWT tokens. Also create a deps.ts file to import/export all required dependencies to simplify importing common modules.

FormalProcess3909 committed:
Completed Tasks 4 & 5 with unit tests. Utilized the utils.test.ts to ensure tasks 4 & 5 unit tests ran correctly. Created a seperate utility file for creating JWT tokens

AlaraBread commented on backend/api/deps.ts:
use the import map instead of re-exporting

AlaraBread replied:
you should delete this file

FormalProcess3909 replied:
File deleted

AlaraBread commented on backend/api/register.ts:
```diff
@@ -0,0 +1,48 @@
+				return;
+			}
+
+			// Hash the password
```
can you salt the password using the username and something unique to our application? something like "team6-resume-analyser" would suffice.

FormalProcess3909 replied:
Created a hashing function for the password and JWT token.

AlaraBread commented on backend/api/userLogin.test.ts:
```diff
@@ -0,0 +1,151 @@
+		email: "user@example.com",
+		username: "user123",
+		password:
+			"debe062ddaaf9f8b06720167c7b65c778c934a89ca89329dcb82ca79d19e17d2", // SHA-256 hash for 'securePassword'
```
it would be clearer and easier to modify if you hashed the string "securePassword" right here instead of hardcoding it.

FormalProcess3909 replied:
Used the hashing function instead of hard coding

AlaraBread commented on backend/api/userLogin.test.ts:
```diff
@@ -0,0 +1,151 @@
+		email: "user@example.com",
+		username: "user123",
+		password:
+			"debe062ddaaf9f8b06720167c7b65c778c934a89ca89329dcb82ca79d19e17d2", // SHA-256 hash for 'securePassword'
```
it would be clearer and easier to modify if you hashed the string right here instead of hardcoding it.

AlaraBread commented on backend/api/userLogin.test.ts:
```diff
@@ -0,0 +1,151 @@
+		email: "user@example.com",
+		username: "user123",
+		password:
+			"debe062ddaaf9f8b06720167c7b65c778c934a89ca89329dcb82ca79d19e17d2", // SHA-256 hash for 'securePassword'
```
it would be clearer and easier to modify if you hashed the string right here instead of hardcoding it.

AlaraBread commented on backend/api/userLogin.ts:
```diff
@@ -0,0 +1,45 @@
+			// Send the token
+			ctx.response.status = 200;
+			ctx.response.body = { token };
+		} catch (error) {
```
you dont need to explicitly define this behavior, oak will send a 500 error automatically. remove the try/catch.

AlaraBread commented on backend/services/createJWTSig.ts:
```diff
@@ -0,0 +1,28 @@
+	alg: "HS256",
+};
+// Secret key for JWT
+const jwtKey = "59c4b48eac7e9ac37c046ba88964870d";
```
Secret key should be in an environment file, not in the code. Please do not use this key anymore. It has already been compromised.

AlaraBread replied:
(and the environment file should never be added to source control)

FormalProcess3909 replied:
Normally I'd never have the key in plain text (the above key is one used for testing to see the effects of creating JWT tokens). I created my own .env file with a JWT key, but I dont know of a way to share the secret key without having to put in a readme what the key is so everyone will have it.

AlaraBread replied:
in prod, we will put this on a server with a key. for testing, you should add instructions in SETUP.md for generating a key and putting it in .env.local

AlaraBread reviewed (changes requested):
please use snake_case for filenames.

also, please implement validating the session and getting the logged in user's username in the job description and resume upload endpoints. you might have to update your branch to get them since they got merged last night.

FormalProcess3909 committed:
Merge branch 'dev-steve' into dev

FormalProcess3909 committed:
Merge pull request #1 from FormalProcess3909/dev

Dev

FormalProcess3909 committed:
Changed filenames to snake_case. Deleted deps.ts file. Created function for hashing. Removed the try/catch with the error status 500.

FormalProcess3909 committed:
Fixed issue with not being able to read .env files with secret JWT key. Need to use [0m[38;5;245mrunning 3 tests from ./api/register_component/user_registration.test.ts[0m
POST /api/register - Successful Registration ... [0m[32mok[0m [0m[38;5;245m(6ms)[0m
POST /api/register - Missing Fields ... [0m[32mok[0m [0m[38;5;245m(0ms)[0m
POST /api/register - Duplicate Email ... [0m[32mok[0m [0m[38;5;245m(3ms)[0m
[0m[38;5;245mrunning 1 test from ./api/user_login_component/user_login.test.ts[0m
POST /api/userLogin - Successful Login ... [0m[32mok[0m [0m[38;5;245m(7ms)[0m
[0m[38;5;245mrunning 1 test from ./hello.test.ts[0m
hello test ... [0m[32mok[0m [0m[38;5;245m(1ms)[0m
[0m[38;5;245mrunning 5 tests from ./in_memory/in_memory.test.ts[0m
Generate session ID ... [0m[32mok[0m [0m[38;5;245m(1ms)[0m
storeData and retrieveData are successful ... [0m[32mok[0m [0m[38;5;245m(0ms)[0m
storeData retrieves null for non-existent sessionId ... [0m[32mok[0m [0m[38;5;245m(0ms)[0m
deleteData removes data correctly ... [0m[32mok[0m [0m[38;5;245m(0ms)[0m
clearAllData removes all session data ... [0m[32mok[0m [0m[38;5;245m(0ms)[0m
[0m[38;5;245mrunning 2 tests from ./upload/job_description_upload.test.ts[0m
Job Description - Valid Input ... [0m[32mok[0m [0m[38;5;245m(9ms)[0m
Job Description - Exceeds Character Limit ... [0m[32mok[0m [0m[38;5;245m(0ms)[0m
[0m[38;5;245mrunning 3 tests from ./upload/resume_upload.test.ts[0m
Valid PDF ... [0m[32mok[0m [0m[38;5;245m(8ms)[0m
Resume Upload - Invalid File Type ... [0m[32mok[0m [0m[38;5;245m(1ms)[0m
Oversized File ... [0m[32mok[0m [0m[38;5;245m(31ms)[0m
[0m[38;5;245mrunning 0 tests from ./util/util.test.ts[0m

[0m[32mok[0m | 15 passed | 0 failed [0m[38;5;245m(448ms)[0m to allow for reading .env files

FormalProcess3909 commented:
Fixed issue with reading .env file. Need to have `deno test --allow-read --allow-env` when running tests for file read permission.

AlaraBread commented:
> Fixed issue with reading .env file. Need to have `deno test --allow-read --allow-env` when running tests for file read permission.

Can you update the github workflow in .github/workflows/deno.yml and the pre-commit hook in hooks/pre-commit with that?

FormalProcess3909 committed:
Finished task 11 along with tests. Created a folder to hold one of the test PDF's and extracted text for testing. Needed to remove the .env from the .gitignore as the name change to .env.local seemed to cause the file to not be found. Updated the command to run tests.

FormalProcess3909 commented:
> > Fixed issue with reading .env file. Need to have `deno test --allow-read --allow-env` when running tests for file read permission.
> 
> Can you update the github workflow in .github/workflows/deno.yml and the pre-commit hook in hooks/pre-commit with that?

just saw this now. Sorry about that. Going to fix that in the githubworkflow

FormalProcess3909 committed:
Ran the deno fmt command. Updated the deno.yml file with the correct testing command.

AlaraBread committed:
parse text from File object instead of file path

FormalProcess3909 committed:
Updated pre-commit file to match deno test in deno.yml

AlaraBread committed:
improve jwt helpers

AlaraBread committed:
use json for login and registration endpoints

FormalProcess3909 committed:
Trying to modify the routes.ts to use middlware to minimize changes to current working file upload api's.

AlaraBread committed:
use token for session data

AlaraBread committed:
add test for jwt

AlaraBread committed:
salt hashes

FormalProcess3909 committed:
Created test file for session_middleware.ts.

AlaraBread committed:
test session middleware
